ui/core/worker_pool.ts

```diff
@@ -155,24 +154,65 @@ export class WorkerPool {
 		return await this.workers[0].isWasmWorker();
 	}
 
-	newProgressHandler(id: string, worker: SimWorker, onProgress: WorkerProgressCallback): (progressData: any) => void {
+	/**
+	 * Start an async request, handling progress and returning the final ProgressMetrics.
+	 * @param requestName
+	 * @param request
+	 * @param id The task id used.
+	 * @param worker
+	 * @param onProgress
+	 * @param totalIterations Used for initial work amount tracking for worker task.
+	 * @returns The final ProgressMetrics.
+	 */
+	private async doAsyncRequest(
+		requestName: SimRequest.raidSimAsync | SimRequest.bulkSimAsync | SimRequest.statWeightsAsync,
+		request: Uint8Array,
+		id: string,
+		worker: SimWorker,
+		onProgress: WorkerProgressCallback,
+		totalIterations: number,
+	): Promise<ProgressMetrics> {
+		try {
+			worker.addSimTaskRunning(id, totalIterations);
+			worker.doApiCall(requestName, request, id);
+			const finalProgress: Promise<ProgressMetrics> = new Promise(resolve => {
+				// Add handler for the progress events
+				worker.addPromiseFunc(
+					this.getProgressName(id),
+					this.newProgressHandler(id, worker, onProgress, pm => resolve(pm)),
+					noop,
+				);
+			});
+			return await finalProgress;
```

Comment:
Is there a need to await this if you return an async/Promise function anyways?

Reply:
Yes, the finally block needs to run after the request is done.